xayn-ai/src/ranker/public.rs

```diff
@@ -0,0 +1,166 @@
+use std::{
+    io::{BufRead, Read},
+    path::Path,
+};
+
+use kpe::Builder as KPEBuilder;
+use layer::io::BinParams;
+use rubert::{AveragePooler, SMBertBuilder};
+
+use crate::{
+    coi::{point::UserInterests, CoiSystem, Configuration as CoiSystemConfiguration},
+    embedding::utils::Embedding,
+    error::Error,
+};
+
+pub struct Ranker(super::Ranker);
+
+impl Ranker {
+    /// Creates a byte representation of the internal state of the ranker.
+    pub fn serialize(&self) -> Result<Vec<u8>, Error> {
+        self.0.serialize()
+    }
+
+    /// Computes the SMBert embedding of the given `sequence`.
+    pub fn compute_smbert(&self, sequence: &str) -> Result<Embedding, Error> {
+        self.0.compute_smbert(sequence)
+    }
+}
+
+pub struct Builder<SV, SM, KV, KM> {
+    smbert: SMBertBuilder<SV, SM>,
+    kpe: KPEBuilder<KV, KM>,
+    user_interests: UserInterests,
+}
+
+impl Default for Builder<(), (), (), ()> {
+    fn default() -> Self {
+        Self {
+            smbert: SMBertBuilder::new((), ()),
+            kpe: KPEBuilder::new((), (), BinParams::default(), BinParams::default()),
+            user_interests: UserInterests::default(),
+        }
+    }
+}
+
+impl<SV, SM, KV, KM> Builder<SV, SM, KV, KM> {
+    /// Sets the serialized state to use.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the state cannot be deserialized.
+    pub fn with_serialized_state(mut self, bytes: impl AsRef<[u8]>) -> Result<Self, Error> {
+        self.user_interests = bincode::deserialize(bytes.as_ref())?;
+        Ok(self)
+    }
+
+    /// Sets the reader of the vocabulary, and bert model used to initialize the SMBert.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the data cannot be deserialized.
+    pub fn with_smbert_from_reader<W, N>(self, vocab: W, bert: N) -> Builder<W, N, KV, KM> {
+        Builder {
+            smbert: SMBertBuilder::new(vocab, bert),
+            kpe: self.kpe,
+            user_interests: self.user_interests,
+        }
+    }
+
+    /// Sets the path of the vocabulary, and bert model used to initialize the SMBert.
+    ///
+    /// # Errors
+    ///
+    /// Fails if a file does not exist or if the data cannot be deserialized.
```

Comment:
```suggestion
    /// Fails if a file cannot be opened.
```

It doesn't fail if it can't deserialize, as it only does read the files when building.

It fails when opening the files fails, which can be for many reasons besides them not existing (mainly permission errors).